docs/design/storage-auto-scale.md

```diff
@@ -0,0 +1,120 @@
+  
+- If the Osd size is less than maxOsdSize(default:8Tib), do vertical scaling by doubling the each osd sizes for that device class.
+
+- If the Osd size is equal to maxOsdSize(default:8Tib), do a horizontal scaling, by adding 1 osd of 8Tib on each storage node, by simply increasing the (count+1) in the device set associated with a device class.   
+
+## Low-Level Design 
+
+Go-routine Scraper:
+
+1) Add a single go-routine to watch periodically(10mins) all the OSDs sizes using Prometheus metrics in the storagecluster namespace.
+    Query: (ceph_osd_metadata * on (ceph_daemon, namespace, managedBy) group_right(device_class,hostname) (ceph_osd_stat_bytes_used / ceph_osd_stat_bytes))
+
+2) With every run calculate different device class highest osd filled percentage.
+   
+        OsdPercentage:
+        
+        ssd               69%
+        nvme              71%
+        ...   
+
+3) Create a sync map with `OsdPercentage`, Send a event to the go channel.
+
+Controller:
+
+1) Create a named controller that watches for [channel generic event](https://book-v1.book.kubebuilder.io/beyond_basics/controller_watches) per device class.
+
+2) If the expansion is not in progress set the status `phase` to `NotStarted` 
+   
+3) If an expansion is in progress(expectedOsdSize!=startOsdSize || expectedOsdCount!=startOsdCount), check the progress and then requeue each 1 minute until the expansion is completed successfully(jump to step 9).
+   
+4) If the highest osd percentage reported in the sync map is more than osdScalingThresholdPercent(70%) means reaching osd nearfull, scaling is needed.
+   
+5) Check if the totalCapacity allows to scale it or not.
+        
+    1) If yes, Update `phase` to `InProgress`  and `lastExpansionStartTime` as `current-time` on the `StorageAutoScaling` CR which need scaling.
+        
+    2) If no, don't reconcile further and change the `phase` to `Failed`
+
+6) If scaling is needed calculate the `expectedOsdSize` and `expectedOsdCount`.
+       
+    1)  If the Osd size is less than maxOsdSize(default:8Tib), do vertical scaling by doubling the each osd sizes for that device class.
+            
+    2)  If the Osd size is equal to maxOsdSize(default:8Tib), do a horizontal scaling, by adding 1 osd of maxOsdSize(default:8Tib) on each `storageDeviceSet`.
+   
+7) Update the status, set the `expectedOsdSize` and `expectedOsdCount` to reflect the new expected value and also set `startOsdSize` and `startOsdCount` with current storagecluster values.
+   
+8) Scale by patching the Storagecluster, with all the device sets update needed at the same time.
+
+9)  Verify and Alert:
+   
+   1) Verify the Storagecluster whether the new osds are added or scaled in size, for all the device sets.
+ 
+   2) If the scaling is successful will raise a `Succeeded` alert and update the status of the `StorageAutoScaling` CR with `lastExpansionCompletionTime` and `phase` and also osd count and size.
```

Comment:
Alerting is based on the phase, right? Let's be clear in this design that we update the phase, and that monitoring will raise the alert based on the phase.
```suggestion
   2) If the scaling is successful will update the status of the `StorageAutoScaling` CR with `lastExpansionCompletionTime` and `phase` and also osd count and size.
```